File: client/android/div/src/main/java/com/yandex/div/core/util/mask/BaseInputMask.kt
package com.yandex.div.core.util.mask

import java.util.regex.PatternSyntaxException
import kotlin.math.min

internal abstract class BaseInputMask(
    initialMaskData: MaskData
) {
    protected var maskData: MaskData = initialMaskData
        private set

    protected val filters = mutableMapOf<Char, Regex>()

    protected lateinit var destructedValue: List<MaskChar>

    var cursorPosition: Int = 0
        protected set

    val rawValue: String
        get() = collectValueRange(0, destructedValue.size - 1)

    val value: String
        get() {
            val stringBuilder = StringBuilder()

            destructedValue.takeWhile { char ->
                return@takeWhile when {
                    char is MaskChar.Static -> {
                        stringBuilder.append(char.char)

                        true
                    }
                    char is MaskChar.Dynamic && char.char != null -> {
                        stringBuilder.append(char.char)

                        true
                    }
                    maskData.alwaysVisible -> {
                        stringBuilder.append((char as MaskChar.Dynamic).placeholder)

                        true
                    }
                    else -> false
                }
            }

            return stringBuilder.toString()
        }

    protected val firstEmptyHolderIndex: Int
        get() {
            return destructedValue.indexOfFirst { maskChar ->
                maskChar is MaskChar.Dynamic && maskChar.char == null
            }.let { if (it != -1) it else destructedValue.size }
        }

    init {
        updateMaskData(initialMaskData)
    }

    open fun updateMaskData(newMaskData: MaskData, restoreValue: Boolean = true) {
        val previousRawValue = if (maskData != newMaskData && restoreValue) rawValue else null

        maskData = newMaskData

        filters.clear()

        maskData.decoding.forEach { maskKey ->
            try {
                maskKey.filter?.let { filters[maskKey.key] = Regex(it) }
            } catch (e: PatternSyntaxException) {
                onException(e)
            }
        }

        destructedValue = maskData.pattern.map { maskChar ->
            val mappingItem = maskData.decoding.firstOrNull { it.key == maskChar }

            if (mappingItem != null)
                MaskChar.Dynamic(null, filters[mappingItem.key], mappingItem.placeholder)
            else
                MaskChar.Static(maskChar)
        }

        if (previousRawValue != null) {
            overrideRawValue(previousRawValue)
        }
    }

    open fun overrideRawValue(newRawValue: String) {
        clearRange(0, destructedValue.size)

        replaceChars(newRawValue, 0)

        cursorPosition = min(cursorPosition, value.length)
    }

    open fun applyChangeFrom(newValue: String, position: Int? = null) {
        val textDiff = TextDiff.build(value, newValue)
            .let {
                if (position != null) {
                    TextDiff(
                        (position - it.added).coerceAtLeast(0),
                        it.added,
                        it.removed
                    )
                } else {
                    it
                }
            }

        val tailStart = replaceBodyTail(textDiff, newValue)

        calculateCursorPosition(textDiff, tailStart)
    }

    protected fun replaceBodyTail(textDiff: TextDiff, newValue: String): Int {
        val body = buildBodySubstring(textDiff, newValue)
        val tail = buildTailSubstring(textDiff)

        cleanup(textDiff)

        val fehi = firstEmptyHolderIndex

        val maxShift = calculateMaxShift(tail, fehi)

        replaceChars(body, fehi, maxShift)

        val tailStart = firstEmptyHolderIndex

        replaceChars(tail, tailStart)

        return tailStart
    }

    private fun buildBodySubstring(textDiff: TextDiff, newValue: String): String {
        return newValue.substring(textDiff.start, textDiff.start + textDiff.added)
    }

    private fun buildTailSubstring(textDiff: TextDiff): String {
        return collectValueRange(textDiff.start + textDiff.removed, destructedValue.size - 1)
    }

    private fun calculateMaxShift(string: String, start: Int): Int {
        return if (filters.size <= 1) {
            var dynamicLeft = 0

            var index = start

            while (index < destructedValue.size) {
                if (destructedValue[index] is MaskChar.Dynamic) dynamicLeft++
                index++
            }

            dynamicLeft - string.length
        } else {
            val initialInsertableSubstring = calculateInsertableSubstring(string, start)

            var index = 0

            while (index < destructedValue.size && initialInsertableSubstring == calculateInsertableSubstring(string, start + index)) {
                index++
            }

            index - 1
        }.coerceAtLeast(0)
    }

    protected fun cleanup(textDiff: TextDiff) {
        if (textDiff.added == 0 && textDiff.removed == 1) {
            var index = textDiff.start

            while (index >= 0) {
                val maskChar = destructedValue[index]

                if (maskChar is MaskChar.Dynamic && maskChar.char != null) {
                    maskChar.char = null

                    break
                } else {
                    index--
                }
            }
        }

        clearRange(textDiff.start, destructedValue.size)
    }

    protected fun clearRange(start: Int, end: Int) {
        var index = start

        while (index < end && index < destructedValue.size) {
            val holder = destructedValue[index]

            if (holder is MaskChar.Dynamic) holder.char = null

            index++
        }
    }

    protected fun calculateCursorPosition(
        textDiff: TextDiff,
        tailStart: Int
    ) {
        val fehi = firstEmptyHolderIndex

        val positionByDiff = if (textDiff.start < fehi) {
            min(firstHolderAfter(tailStart), value.length)
        } else {
            fehi
        }

        cursorPosition = positionByDiff
    }

    protected fun calculateInsertableSubstring(substring: String, start: Int): String {
        val charsCanBeInsertedStringBuilder = StringBuilder()

        var index = start

        val moveToAndGetNextHolderFilter: () -> Regex? = {
            while (index < destructedValue.size && destructedValue[index] !is MaskChar.Dynamic) {
                index++
            }

            (destructedValue.getOrNull(index) as? MaskChar.Dynamic)?.filter
        }

        substring.forEach { char ->
            val maskCharFilter = moveToAndGetNextHolderFilter()

            when {
                maskCharFilter == null -> Unit
                maskCharFilter.matches(char.toString()) -> {
                    charsCanBeInsertedStringBuilder.append(char)

                    index++
                }
            }
        }

        return charsCanBeInsertedStringBuilder.toString()
    }

    protected fun collectValueRange(start: Int, end: Int): String {
        val tailStringBuilder = StringBuilder()

        var index = start

        while (index <= end) {
            val maskChar = destructedValue[index]

            if (maskChar is MaskChar.Dynamic) {
                if (maskChar.char != null) {
                    tailStringBuilder.append(maskChar.char)
                }
            }

            index++
        }

        return tailStringBuilder.toString()
    }

    protected fun replaceChars(substring: String, start: Int, count: Int? = null) {
        val trimmedSubstring = calculateInsertableSubstring(substring, start).let {
            if (count != null) it.take(count) else it
        }

        var index = start

        var insertableCharIndex = 0

        while (index < destructedValue.size && insertableCharIndex < trimmedSubstring.length) {
            val maskChar = destructedValue[index]
            val char = trimmedSubstring[insertableCharIndex]

            if (maskChar is MaskChar.Dynamic) {
                maskChar.char = char

                insertableCharIndex++
            }

            index++
        }
    }

    protected fun firstHolderAfter(start: Int): Int {
        var index = start

        while (index < destructedValue.size) {
            val holder = destructedValue[index]

            if (holder is MaskChar.Dynamic) {
                break
            } else {
                index++
            }
        }

        return index
    }

    abstract fun onException(exception: Exception)

    class MaskKey(
        val key: Char,
        val filter: String?,
        val placeholder: Char
    )

    data class MaskData(
        val pattern: String,
        val decoding: List<MaskKey>,
        val alwaysVisible: Boolean
    )

    sealed class MaskChar {
        data class Static(val char: Char) : MaskChar()

        data class Dynamic(var char: Char?, val filter: Regex?, val placeholder: Char) : MaskChar()
    }
}

File: client/web/divkit/src/utils/mask/baseInputMask.ts
// Replicate from Android: client/android/div/src/main/java/com/yandex/div/core/util/mask/BaseInputMask.kt

import type { WrappedError } from '../wrapError';
import { wrapError } from '../wrapError';
import { type TextDiff, textDiff } from '../textDiff';

export interface MaskKey {
    key: string;
    filter?: string | undefined;
    placeholder: string;
}

export interface MaskData {
    pattern: string;
    decoding: MaskKey[];
    alwaysVisible: boolean;
}

interface MaskChar {
    char: string | null;
}

export class MaskCharStatic implements MaskChar {
    constructor(public char: string) {}
}

export class MaskCharDynamic implements MaskChar {
    constructor(public char: string | null, public filter: RegExp, public placeholder: string) {}
}

export abstract class BaseInputMask {
    protected maskData: MaskData;
    protected filters: Map<string, RegExp> = new Map();
    protected destructedValue: MaskChar[] = [];
    protected cursorPos = 0;

    constructor(initialMaskData: MaskData) {
        this.maskData = initialMaskData;
        this.updateMaskData(initialMaskData);
    }

    get cursorPosition(): number {
        return this.cursorPos;
    }

    get rawValue(): string {
        return this.collectValueRange(0, this.destructedValue.length - 1);
    }

    get value(): string {
        let str = '';

        for (let i = 0; i < this.destructedValue.length; ++i) {
            const char = this.destructedValue[i];

            if (char instanceof MaskCharStatic) {
                str += char.char;
            } else if (char instanceof MaskCharDynamic) {
                if (char.char) {
                    str += char.char;
                } else if (this.maskData.alwaysVisible) {
                    str += char.placeholder;
                } else {
                    break;
                }
            }
        }

        return str;
    }

    protected firstEmptyHolderIndex(): number {
        const index = this.destructedValue.findIndex(maskChar => {
            return maskChar instanceof MaskCharDynamic && !maskChar.char;
        });

        if (index !== -1) {
            return index;
        }

        return this.destructedValue.length;
    }

    abstract onException(exception: WrappedError): void;

    updateMaskData(newMaskData: MaskData, restoreValue = true): void {
        const previousRawValue = this.maskData !== newMaskData && restoreValue ? this.rawValue : null;

        this.filters = new Map();
        this.maskData = newMaskData;

        this.maskData.decoding.forEach(maskKey => {
            if (maskKey.filter) {
                try {
                    const regexp = new RegExp(maskKey.filter);
                    this.filters.set(maskKey.key, regexp);
                } catch (err) {
                    this.onException(wrapError(err as Error, {
                        level: 'error',
                        additional: {
                            key: maskKey.key
                        }
                    }));
                }
            }
        });

        this.destructedValue = this.maskData.pattern.split('').map(maskChar => {
            const mappingItem = this.maskData.decoding.find(it => it.key === maskChar);

            if (mappingItem) {
                return new MaskCharDynamic(
                    null,
                    this.filters.get(mappingItem.key) as RegExp,
                    mappingItem.placeholder
                );
            }

            return new MaskCharStatic(maskChar);
        });

        if (previousRawValue !== null) {
            this.overrideRawValue(previousRawValue);
        }
    }

    overrideRawValue(newRawValue: string): void {
        this.clearRange(0, this.destructedValue.length);

        this.replaceChars(newRawValue, 0);

        this.cursorPos = Math.min(this.cursorPos, this.value.length);
    }

    applyChangeFrom(newValue: string, position?: number): void {
        const diff = textDiff(this.value, newValue);

        if (position !== undefined) {
            diff.start = Math.max(0, position - diff.added);
        }

        const tailStart = this.replaceBodyTail(diff, newValue);

        this.calculateCursorPosition(diff, tailStart);
    }

    protected replaceBodyTail(diff: TextDiff, newValue: string): number {
        const body = this.buildBodySubstring(diff, newValue);
        const tail = this.buildTailSubstring(diff);

        this.cleanup(diff);

        const fehi = this.firstEmptyHolderIndex();

        const maxShift = tail ? this.calculateMaxShift(tail, fehi) : undefined;

        this.replaceChars(body, fehi, maxShift);

        const tailStart = this.firstEmptyHolderIndex();

        this.replaceChars(tail, tailStart);

        return tailStart;
    }

    protected buildBodySubstring(diff: TextDiff, newValue: string): string {
        return newValue.substring(diff.start, diff.start + diff.added);
    }

    protected buildTailSubstring(diff: TextDiff): string {
        return this.collectValueRange(
            diff.start + diff.removed,
            this.destructedValue.length - 1
        );
    }

    protected calculateMaxShift(str: string, start: number): number {
        if (this.filters.size <= 1) {
            let dynamicLeft = 0;
            let index = start;

            while (index < this.destructedValue.length) {
                if (this.destructedValue[index] instanceof MaskCharDynamic) {
                    ++dynamicLeft;
                }
                ++index;
            }

            return Math.max(0, dynamicLeft - str.length);
        }

        const initialInsertableSubstring = this.calculateInsertableSubstring(str, start);

        let index = 0;

        while (
            index < this.destructedValue.length &&
                initialInsertableSubstring === this.calculateInsertableSubstring(str, start + index)
        ) {
            ++index;
        }

        return Math.max(0, index - 1);
    }

    protected cleanup(diff: TextDiff): void {
        if (diff.added === 0 && diff.removed === 1) {
            let index = diff.start;

            while (index >= 0) {
                const maskChar = this.destructedValue[index];

                if (maskChar instanceof MaskCharDynamic && maskChar.char !== null) {
                    maskChar.char = null;

                    break;
                } else {
                    --index;
                }
            }
        }

        this.clearRange(diff.start, this.destructedValue.length);
    }

    protected clearRange(start: number, end: number): void {
        let index = start;

        while (index < end && index < this.destructedValue.length) {
            const holder = this.destructedValue[index];

            if (holder instanceof MaskCharDynamic) {
                holder.char = null;
            }

            ++index;
        }
    }

    protected calculateCursorPosition(diff: TextDiff, tailStart: number) {
        const fehi = this.firstEmptyHolderIndex();

        let positionByDiff: number;
        if (diff.start < fehi) {
            positionByDiff = Math.min(this.firstHolderAfter(tailStart), this.value.length);
        } else {
            positionByDiff = fehi;
        }

        this.cursorPos = positionByDiff;
    }

    protected calculateInsertableSubstring(substring: string, start: number): string {
        let str = '';

        let index = start;

        const moveToAndGetNextHolderFilter: () => RegExp | undefined = () => {
            while (index < this.destructedValue.length && !(this.destructedValue[index] instanceof MaskCharDynamic)) {
                ++index;
            }

            return (this.destructedValue[index] as MaskCharDynamic)?.filter;
        };

        substring.split('').forEach(char => {
            const maskCharFilter = moveToAndGetNextHolderFilter();

            if (maskCharFilter?.test(char)) {
                str += char;
                ++index;
            }
        });

        return str;
    }

    protected collectValueRange(start: number, end: number): string {
        let str = '';

        let index = start;

        while (index <= end) {
            const maskChar = this.destructedValue[index];

            if (maskChar instanceof MaskCharDynamic && maskChar.char !== null) {
                str += maskChar.char;
            }

            ++index;
        }

        return str;
    }

    protected replaceChars(substring: string, start: number, count?: number): void {
        let trimmedSubstring = this.calculateInsertableSubstring(substring, start);
        if (count !== undefined) {
            trimmedSubstring = trimmedSubstring.substring(0, count);
        }

        let index = start;

        let insertableCharIndex = 0;

        while (index < this.destructedValue.length && insertableCharIndex < trimmedSubstring.length) {
            const maskChar = this.destructedValue[index];
            const char = trimmedSubstring[insertableCharIndex];

            if (maskChar instanceof MaskCharDynamic) {
                maskChar.char = char;

                ++insertableCharIndex;
            }

            ++index;
        }
    }

    protected firstHolderAfter(start: number): number {
        let index = start;

        while (index < this.destructedValue.length) {
            const holder = this.destructedValue[index];

            if (holder instanceof MaskCharDynamic) {
                break;
            } else {
                ++index;
            }
        }

        return index;
    }
}
